Subfeature/6 temporal alignment

drpjm opened the pull request:
<!-- Thanks for the contribution, this is awesome. -->

# PR Details
## Description

This feature builds out the core data objects and methods to perform temporal alignment of detected objects.

## Related Issue

Partially addresses #8 

## Motivation and Context

This code adds features that are required for our object tracking framework for the host vehicle's detected objects.

## How Has This Been Tested?

CTRV model, DetectedObjects, and Unscented Transforms were all tested with best practice C++ tests.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

adamlm committed:
Install Boost.Container in devcontainer

The project now depends on Boost.Container, but the devcontainer does
not install in when being built.

.devcontainer/Docker
* add libboost-container-dev to package installation

adamlm committed:
Add CtrvStateCovariance and VehicleObject

The CtrvStateCovariance is a type alias for an Eigen matrix. It would be nice
to have a units-based type, but this is currently a hard limitation for all
open source linear algebra libraries. The VehicleObject is a type alias for
a DetectedObject with a CTRV motion model.

dependencies.cmake
* add Eigen3

include/cooperative_perception/ctrv_model.hpp
* add CtrvStateCovariance type alias

include/cooperative_perception/detected_object.hpp
* add VehicleObject type alias

src/CMakeLists.txt
* link to Eigen3

adamlm committed:
Add initial implementation for temporal alignment

The implementation uses a functional programming style due to the templated
DetectedObject class.

include/cooperative_perception/temporal_alignment.hpp
* add state propagator lambda that will propagate the state for the passed in
  object. This uses the visitor pattern
* add function to temporally align a single detected object
* add function to temporally align a list of detected objects

include/cooperative_perception/utils.hpp
* add Visitor class

adamlm committed:
Add sigma point sampling for unscented transform

To propagate the static covariance matrix, we need to approximate it
from a sampling of state vectors. This commit adds the sample generation
function and supporting code.

.devcontainer/Dockerfile
* upgrade to Eigen 3.4 because it provides modern features like
  initializer lists and range based for loops

.vscode/settings.json
* switch to external clang-format and clang-tidy because the ones
  integrated into VS Code seem to have some behavioral issues

dependencies.cmake
* specify Eigen 3.4 version

include/cooperative_perception/ctrv_model.hpp
* add addition and subtraction (assignment) operators because they are
  necessary for sampling state space points in the unscented transform
* add equality operator because it is necessary for unordered
  containers. Besides that, this function will probably rarely be used.
* add hash_value function because it is needed for the boost::hash
  function
* add almostEqual function to "properly" compare equality for the state
  variables. It has a tight 4 ULP tolerance that could probably be moved
  to a function parameter
* add roundToDecimalPlaces function that will round the state variables
  to the specified number of decimal places. This was added because some
  of the hand calculated unit test states did not have the precision
  that could be done programmatically and it was causing the tests to
  fail

include/cooperative_perception/unscented_transform.hpp
* add function declaration for sampling state vectors

src/CMakeLists.txt
* add unscented_transform.cpp to sources list

src/ctrv_model.cpp
* add hash_value function implementation

src/unscented_transform.cpp
* add sampleStateDistribution function implementation

test/CMakeLists.txt
* add unit test for unscented transform to target sources

test/test_ctrv_model.cpp
* add unit test for almostEqual

test/test_unscented_transform.cpp
* add unit test for sampleStateDistribution

adamlm committed:
Add namespace alias to unit test files

The unit test files were using the cooperative_perception namespace,
which reduced readability. The namespace alias cp is replacing the full
namespace. This shouldn't be an issue because the abbreviation should be
clear from context.

test/test_ctrv_model.cpp
* replace full cooperative_perception namespace with cp alias

test/test_detected_object.cpp
* replace full cooperative_perception namespace with cp alias

test/test_unscented_transform.cpp
* replace full cooperative_perception namespace with cp alias

adamlm committed:
Clean up CTRV tests

With the new almostEqual and roundToDecimalPlaces utility functions, the
old manual way of testing CtrvState equality can be replaced.

test/test_ctrv_model.cpp
* replace manual equality testing for CtrvStates with utility functions
* add constexpr where possible

adamlm committed:
Move CtrvState add/subtract assignment operators

The add and subtract assignment operators for the CtrvState struct do
not need to access private members because it is a POD. Therefore, the
functions can be made free functions.

include/cooperative_perception/ctrv_model.hpp
* convert add assignment operator to free function
* convert subtract assignment operator to free function

adamlm committed:
Update VS Code workspace file associations

The workspace settings allowed file associations to be automatically
added to the workspace settings, but this meant any standard header file
included in the project would be added to this list (since header files
have no extension). This is unnecessary, so automatic file associations
have been disabled. Existing associations have also been removed because
they kind of snuck in.

.vscode/settings.json
* disable automatic file associations
* remove existing file associations

adamlm committed:
Add std::hash specialization for CtrvState

A hash mechanism and equality operator must be provided to use
user-defined types in unordered containers. Previously, the boost::hash
specialization for CtrvState was used, but this required passing it as a
template argument whenever CtrvState was used in an unordered container.
This was a bit annoying, so it was replaced with an alternative where
the std::hash specialization for CtrvState is injected into the std
namespace.

include/cooperative_perception/ctrv_model.hpp
* remove hash_value function because it is no longer needed
* add std::hash specialization for CtrvState

include/cooperative_perception/unscented_transform.hpp
* remove boost::hash template argument

src/ctrv_model.cpp
* remove hash_value function implementation

src/unscented_transform.cpp
* remove boost::hash template arguments

test/test_unscented_transform.cpp
* remove boost::hash template arguments

adamlm committed:
Rename roundToDecimalPlaces function

The function name was kind of awkward to read and was slightly different
from how people normally speak about decimal places.

include/cooperative_perception/ctrv_model.hpp
* rename roundToDecimalPlaces to roundToDecimalPlace

test/test_ctrv_model.cpp
* replace roundToDecimalPlaces with roundToDecimalPlace

test/test_unscented_transform.cpp
* replace roundToDecimalPlaces with roundToDecimalPlace

adamlm committed:
Add CTRV process noise struct

The CTRV process noise was passed to the nextState function via separate
function arguments, but this will make generalizing the function
signature difficult. The new CtrvProcessNoise struct will help unify the
signature.

include/cooperative_perception/ctrv_model.hpp
* add CtrvProcessNoise struct
* update stochastic nextState signature to use CtrvProcessNoise struct

src/ctrv_model.hpp
* update stochastic nextState signature and implementation to use CtrvProcessNoise struct

test/test_ctrv_model.cpp
* update stochastic nextState unit test to use CtrvProcessNoise struct

adamlm committed:
Clean up CTRV nextState implementations

The previous implementations were overly verbose, diminishing
readability.

src/ctrv_model.cpp
* clean up deterministic nextState implementation
* clean up stochastic nextState implementation

adamlm committed:
Generalize state space sampling for UT

The unscented transform implementation was previously specific to the
CTRV model, but now it should work for any model assuming they provide
the expected interfaces.

include/cooperative_perception/ctrv_model.hpp
* add static function to create CtrvState from an Eigen vector.
  Generally speaking, I don't like doing this because it might encourage
  users to store state data in dimensionless Eigen vectors. However, I
  think the utility we gain from it, especially for state space
  sampling, outweighs the risks.

include/cooperative_perception/unscented_transform.hpp
* add templated version of sampleStateDistribution function. This
  replaces the CTRV specific implementation
* remove sampleStateDistribution function declaration for CTRV model

src/CMakeLists.txt
* remove unscented_transform.cpp from target sources because it was
  deleted

src/unscented_transform.cpp
* remove file because it only had the CTRV model implementation for
  sampleStateDistribution. The templated replacement is in the
  unscented_transform.hpp header file

adamlm committed:
Add augmented state sigma point generation

A new AugmentedState container has been added that allows the unscented
transform to be used in an augmented state context. Mathematically, the
augmented state vector concatenates the system's state vector and its
process noise vector. Programmatically, the AugmentedState struct
contains a state field and a process_noise field. This prevents
duplicate code.

include/cooperative_perception/augmented_state.hpp
* add AugmentedState implementation
* add boilerplate add/subtract (and assign versions) and equality
  comparison operator overloads
* add std::hash specialization so it can be used with unordered
  containers
* add almostEqual implementation
* add roundToDecimalPlace implementation

include/cooperative_perception/ctrv_model.hpp
* add static member variable to convey how many state variables are in
  CtrvState. This is needed for when the CtrvState is consumed in an
  AugmentedState
* add static member variable to convey how many variables are in
  CtrvProcessNoise
* add fromEigenVector implementation for CtrvProcessNoise
* add add/subtract (with assignment) and equality operator overloads for
  CtrvProcessNoise
* move ULP tolerance in almostEqual functions to parameters so they can
  be adjusted if needed
* add std::hash implementation for CtrvProcessNoise

test/test_unscented_transform.cpp
* add unit test for AgumentedState sigma point generation with the CTRV
  motion model

adamlm committed:
Add copyright header to all code files

Add Apache-2.0 license header to all code files.

adamlm commented on include/cooperative_perception/detected_object.hpp:
```diff
@@ -16,7 +33,9 @@ struct DetectedObject
+
+using DetectedObjectType = std::variant<VehicleObject>;
 
 using DetectedObjectList = boost::container::static_vector<DetectedObjectType, 200>;
```
The `200` is a magic number and should be moved to a compile time constant.

drpjm replied:
Maybe `static constexpr auto kMaxDetectedObjects{ 200 };` within  `detected_object.hpp`?

adamlm replied:
Yes, so the new code would look like

```cpp
namespace cooperative_perception {
// ...
inline constexpr auto kMaxDetectedObjects{ 200U };
using DetectedObjectList = boost::container::static_vector<DetectedObjectType, kMaxDetectedObjects>;
// ...
}
```

drpjm replied:
Cool - I updated on my local copy before I saw this to use:
`static constexpr auto kMaxDetectedObjects {200};`
I was just copying the way I saw you setup the constant for `CtrvModel` in its header. Why would we use `inline` vs `static`?

adamlm replied:
We use `static constexpr` in the `CtrvState` because `kNumVars` should not be an instance variable. All instances of `CtrvState` have five elements.

We use `inline constexpr` in the `namespace` scope because `kMaxDetectedObjects` could have multiple definitions among translation units. The `inline` specifier will tell the linker to merge all definitions into one. I need to review `inline` some more for a better distinction, but here's a short [article](https://quuxplusone.github.io/blog/2022/07/08/inline-constexpr/) about how it's useful in this context.

adamlm commented on include/cooperative_perception/ctrv_model.hpp:
```diff
@@ -1,21 +1,141 @@
+  return lhs;
+}
+
+using CtrvStateCovariance = Eigen::Matrix<float, 5, 5>;
```
The matrix dimensions should use `CtrvState::kNumVars` to avoid size mismatches.

drpjm replied:
Would this be solved by reusing `static constexpr auto kNumVars{ 5 };` from further up the file? So:
`Eigen::Matrix<float, kNumVars, kNumVars>`

adamlm replied:
Yes, but the syntax is a little off. It would look like

```cpp
namespace cooperative_perception {
// ...
using CtrvStateCovariance = Eigen::Matrix<float, CtrvState::kNumVars, CtrvState::kNumVars>;
// ...
}

drpjm replied:
Makes sense - I will make changes tomorrow

adamlm reviewed (changes requested):
There are a few minor things that need to be addressed. See specific comments.

adamlm committed:
Remove magic numbers

The capacity limit on detected objects and the CtrvStateCovariance matrix
size both used magic numbers. They have been moved to constexpr
variables to provide more context.

include/cooperative_perception/ctrv_model.hpp
* replace CtrvStateCovariance matrix dimensions with CtrvState::kNumVars

include/cooperative_perception/detected_object.hpp
* replace detected object capacity limit magic number with constexpr
  variable
